validator/tests/unit3/test_scheduler/tests.py

```diff
@@ -299,104 +297,788 @@ def test_valid_batch_invalid_batch(self):
 
 
 class TestPredecessorTree(unittest.TestCase):
-
-    def test_predecessor_tree(self):
-        """Tests basic functionality of the scheduler's radix tree.
+    '''
+    With an empty tree initialized in setUp, the predecessor tree
+    tests generally follow this pattern (repeated several times):
+
+        1) Add some readers or writers.
+        2) Assert the readers, writers, and children at all addresses
+           in the tree (using assert_rwc_at_addresses). Possibly also
+           assert (using assert_no_nodes_at_addresses) that nodes don't
+           exist at certain addresses (this is normally done after
+           setting a writer).
+        3) Assert the total count of readers and writers in the tree
+           (using assert_rw_count). This ensures that nothing is in the
+           tree that shouldn't be there.
+        4) Assert the read and write predecessors for various addresses
+           (using assert_rw_preds_at_addresses).
+
+    Although the default token size for the predecessor tree is 2,
+    1 is used for most tests because it makes for more natural examples.
+
+    self.verbose is a flag used for printing. It's turned off by default,
+    but it can be turned on for indivudual tests. If the flag is set,
+    assert_rwc_at_addresses will print the entire tree. It's a handy
+    debugging tool.
+
+    '''
+
+    def setUp(self):
+        self.verbose = 0
+        self.tree = PredecessorTree(token_size=1)
+
+    def tearDown(self):
+        self.tree = None
+
+    def test_common_initial_segments(self):
+        '''Tests basic predecessor tree functions
```

Comment:
I feel like the above class-level comment does a good job explaining the general approach, but that I'm left at the per-statement level wondering why the specific readers/writers/assertions are made.  I feel like we need some ASCII way to visualize the tree, and then we could have comments "the tree is now expected to looks like this: ...", which would then explain the result and why we are checking the addresses as we are.  Wondering if we can use the verbose output to capture the tree representations and then embed that as comments.